cmd/snap-update-ns/change.go

```diff
@@ -121,6 +120,25 @@ func (c *Change) createPath(path string, pokeHoles bool, as *Assumptions) ([]*Ch
 		err = MkfileAll(path, mode, uid, gid, rs)
 	case "symlink":
 		err = MksymlinkAll(path, mode, uid, gid, c.Entry.XSnapdSymlink(), rs)
+	case "ensure-dir":
+		uid = sysGetuid()
+		gid = sysGetgid()
+		if uid == 0 {
+			logger.Noticef("WARNING: cannot perform ensure-dir mount for target %q, not supported for the root user", path)
+		} else {
+			// https://specifications.freedesktop.org/basedir-spec/basedir-spec-latest.html
+			// Mode hints cannot be used here because it does not support specifying all
+			// directory paths within a given directory.
+			mode = 0700
+
+			err = MkdirAllWithin(path, c.Entry.XSnapdMustExistDir(), mode, uid, gid, rs)
```

Comment:
@ZeyadYasser, thanks for asking about not using restrictions. Your question made me reconsider and I decided to use restrictions. Note that modeHints could not be used, because it does not allow glob style expression that includes child dirs.